src/gt4py/caching.py

```diff
@@ -0,0 +1,412 @@
+            gt4py.utils.make_dir(str(cache_root), is_cache=True)
+
+        return cache_root
+
+    @property
+    def backend_root_path(self) -> pathlib.Path:
+        """Get the caching base path for the current build."""
+        cpython_id = "py{version.major}{version.minor}_{api_version}".format(
+            version=sys.version_info, api_version=sys.api_version
+        )
+        return self.root_path / cpython_id / gt4py.utils.slugify(self.builder.backend.name)
+
+    @property
+    def cache_info_path(self) -> Optional[pathlib.Path]:
+        """Get the cache info file path from the stencil module path."""
+        return self.builder.module_path.parent / f"{self.builder.module_path.stem}.cacheinfo"
+
+    def generate_cache_info(self) -> Dict[str, Any]:
+        """Generate the cache info dict, reading additional info from the backend."""
+        return {
+            "backend": self.builder.backend.name,
+            "stencil_name": self.builder.stencil_id.qualified_name,
+            "stencil_version": self.builder.stencil_id.version,
+            "module_shash": gt4py.utils.shash(self.builder.stencil_source),
+            **self.builder.backend.extra_cache_info,
+        }
+
+    def update_cache_info(self) -> None:
+        """Write the current cache info to the cache info file."""
+        if not self.cache_info_path:
+            return
+        cache_info = self.generate_cache_info()
+        self.cache_info_path.parent.mkdir(parents=True, exist_ok=True)
+        with self.cache_info_path.open("wb") as cache_info_file:
+            pickle.dump(cache_info, cache_info_file)
+
+    def validate_cache_info(
+        self, cache_info: Dict[str, Any], *, validate_hash: bool, catch_exceptions: bool = True
+    ) -> bool:
+        """Validate the cache info."""
+        result = True
+        try:
+            cache_info_ns = types.SimpleNamespace(**cache_info)
+            validate_extra = self.builder.backend.extra_cache_validate
+            source = self.builder.module_path.read_text()
+            module_shash = gt4py.utils.shash(source)
+
+            if validate_hash:
+                result = (
+                    cache_info_ns.backend == self.builder.backend.name
+                    and cache_info_ns.stencil_name == self.builder.stencil_id.qualified_name
+                    and cache_info_ns.stencil_version == self.builder.stencil_id.version
+                    and cache_info_ns.module_shash == module_shash
+                )
+                if validate_extra:
+                    result &= all(
+                        [cache_info[key] == validate_extra[key] for key in validate_extra]
+                    )
+        except Exception as err:
+            if not catch_exceptions:
+                raise err
+            result = False
+
+        return result
+
+    def check_cache_info(self, *, validate_hash: bool, catch_exceptions: bool = True) -> bool:
+        """Check if the cache can be read and is consistent."""
```

Comment:
I cannot speak for python, but in general I'd say repeating a comment duplication is worse than code duplication.